server/src/main/java/io/crate/planner/optimizer/rule/OptimizeJoinWithLookup.java

```diff
@@ -0,0 +1,247 @@
+ * additional information regarding copyright ownership.  Crate licenses
+ * this file to you under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial agreement.
+ */
+
+package io.crate.planner.optimizer.rule;
+
+import static io.crate.common.collections.Iterables.getOnlyElement;
+import static io.crate.planner.optimizer.matcher.Pattern.typeOf;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.UnaryOperator;
+
+import io.crate.analyze.relations.AbstractTableRelation;
+import io.crate.analyze.relations.FieldResolver;
+import io.crate.common.collections.Maps;
+import io.crate.expression.symbol.ScopedSymbol;
+import io.crate.expression.symbol.Symbol;
+import io.crate.metadata.NodeContext;
+import io.crate.metadata.RelationName;
+import io.crate.metadata.TransactionContext;
+import io.crate.planner.consumer.RelationNameCollector;
+import io.crate.planner.operators.Collect;
+import io.crate.planner.operators.EquiJoinDetector;
+import io.crate.planner.operators.JoinConditionSymbolsExtractor;
+import io.crate.planner.operators.JoinPlan;
+import io.crate.planner.operators.LogicalPlan;
+import io.crate.planner.operators.LogicalPlanVisitor;
+import io.crate.planner.operators.LookupJoin;
+import io.crate.planner.operators.Rename;
+import io.crate.planner.optimizer.Rule;
+import io.crate.planner.optimizer.costs.PlanStats;
+import io.crate.planner.optimizer.iterative.GroupReference;
+import io.crate.planner.optimizer.matcher.Captures;
+import io.crate.planner.optimizer.matcher.Pattern;
+import io.crate.sql.tree.JoinType;
+
+/**
+ * This rule optimizes equi-inner-joins with a lookup-join operator if there is
+ * a large imbalance between the two relations joined.
+ */
+public class OptimizeJoinWithLookup implements Rule<JoinPlan> {
+
+    private final Pattern<JoinPlan> pattern = typeOf(JoinPlan.class).with(j ->
+        j.isLookUpJoinRuleApplied() == false &&
+            // only non-nested inner-equi-joins
+            j.joinType() == JoinType.INNER &&
+            j.joinCondition() != null &&
+            EquiJoinDetector.isEquiJoin(j.joinCondition()) &&
+            j.lhs().getRelationNames().size() == 1 &&
+            j.rhs().getRelationNames().size() == 1
```

Comment:
Could you add a comment explaining why the relation name size must be 1?

Reply:
The join should not be nested. It's already in the comment. Maybe it was not clear.